drivers/SAI.cpp

```diff
@@ -0,0 +1,70 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2013 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "drivers/SAI.h"
+#include "platform/mbed_critical.h"
+
+#if DEVICE_SAI
+
+namespace mbed {
+
+SAI::SAI(const sai_format_t *fmt, bool is_input, uint32_t master_clock, bool internal_mclk) : _sai(), _mutex(), _is_input(is_input) {
+    // No lock needed in the constructor
+    sai_init_t init = {};
+
+    init.mclk = SAI_MCLK;
+    if (_is_input) {
+        init.sd = SAI_RX_SD;
+        init.bclk = SAI_RX_BCLK;
+        init.wclk = SAI_RX_WCLK;
+    } else {
+        init.sd = SAI_TX_SD;
+        init.bclk = SAI_TX_BCLK;
+        init.wclk = SAI_TX_WCLK;
+    }
+
+    init.is_receiver = is_input;
+    init.sample_rate = 16000;
+    init.mclk_source = SAI_CLOCK_SOURCE_Internal;
+    init.input_mclk_frequency = 0; // 0 means find it by yourself.
+    init.output_mclk_frequency = 384 * init.sample_rate;
```

Comment:
You have used a magic number (`384`) to define the output master clock freq. Is this value specific for this target? Why it has been used?

I'm asking since I'm wandering how it should be defined in the test. I found information that output master clock freq should be defined based on sampling freq, number of channels(1-mono, 2-stereo) and number of transmitted bits per channel as follows:
`master clock freq = number of channels * number of transmitted bits per channel * sampling freq`
But I can't see how this translates to `384`.

Reply:
It was actually the value used by Freescale's demo in the SDK.

It is defined there as `OVER_SAMPLE_RATE` in `SDK_2.3.0_FRDM-K66F/boards/frdmk66f/demo_apps/sai/`